### PR TITLE
Added player permission script commands

### PR DESCRIPTION
**Pull request by Daegaladh**
* **Addressed Issue(s)**: N/A

* **Server Mode**: Both

* **Description of Pull Request**:  Adds script commands to check and temporarily add or remove player permissions.

### CODE REVIEW

File: doc/script_commands.txt [hunk 1]
```diff
@@ -6660,6 +6660,46 @@ Example:
 
 ---------------------------------------
 
+*checkpermission(<permission>{,<char_id>});
```
Comment (Daegaladh) [inline doc/script_commands.txt:6663]:
```suggestion
*checkperm(<permission>{,<char_id>});
```

File: doc/script_commands.txt [hunk 2]
```diff
@@ -6660,6 +6660,46 @@ Example:
...
+
+This command will return true if the attached character has the specified permission, false otherwise.
+If <char_id> is given, it will check the permission for that character instead.
+
+A full list of the player permissions can be found in 'src/map/script_constants.hpp', with the 'PC_PERM' prefix,
+and documentation can be found in 'doc/permissions.txt'.
+
+
+Example:
+	if (checkpermission(PC_PERM_TRADE)) {
```
Comment (Daegaladh) [inline doc/script_commands.txt:6673]:
```suggestion
	if (checkperm(PC_PERM_TRADE)) {
```

File: src/map/script.cpp [hunk 3]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
 	return SCRIPT_CMD_SUCCESS;
 }
 
+/**
+ * Checks if the character has that permission.
+ * checkpermission(<permission>{,<char_id>}) -> <bool>
```
Comment (Daegaladh) [inline src/map/script.cpp:27345]:
```suggestion
 * checkperm(<permission>{,<char_id>}) -> <bool>
```

File: src/map/script.cpp [hunk 4]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+/**
+ * Adds or removes a permission from the character
+ * addperm(<permission>{,<char_id>})
+ * removeperm(<permission>{,<char_id>})
+ */
+BUILDIN_FUNC(addperm)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(4, sd))
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27371]: 3

File: src/map/script.cpp [hunk 5]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+		return SCRIPT_CMD_FAILURE;
+
+	e_pc_permission permission;
+	if (script_hasdata(st, 2))
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27375]: 2 is defined as mandatory in the definition, so this check is useless

File: src/map/script.cpp [hunk 6]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+		permission = static_cast<e_pc_permission>(script_getnum(st, 2));
+	else
+		return SCRIPT_CMD_FAILURE;
+
+	bool add = (!strcmp(script_getfuncname(st), "addperm")) ? true : false;
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27380]: Not really a friend of this, if it is not really necessary

File: src/map/script.cpp [hunk 7]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+BUILDIN_FUNC(addperm)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(4, sd))
+		return SCRIPT_CMD_FAILURE;
+
+	e_pc_permission permission;
+	if (script_hasdata(st, 2))
+		permission = static_cast<e_pc_permission>(script_getnum(st, 2));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27376]: Validation of range is missing

File: src/map/script.cpp [hunk 8]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+ * Checks if the character has that permission.
+ * checkperm(<permission>{,<char_id>}) -> <bool>
+ */
+BUILDIN_FUNC(checkperm)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(3, sd))
+		return SCRIPT_CMD_FAILURE;
+
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27353]: Please store the permission in a variable.
Then verify the range. (Out of range => script cmd failure)
Then cast it.

File: src/map/script.cpp [hunk 9]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+	if (pc_has_permission(sd, static_cast<e_pc_permission>(script_getnum(st, 2))))
+		script_pushint(st, true);
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27355]: You can directly push the result of pc_has_permission

**Review by Lemongrass3110 — COMMENTED** [review]
In general I would suggest to rename all the commands to permission_xxx instead.

File: src/map/script.cpp [hunk 10]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+ * checkperm(<permission>{,<char_id>}) -> <bool>
+ */
+BUILDIN_FUNC(checkperm)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(3, sd))
+		return SCRIPT_CMD_FAILURE;
+
+	if (pc_has_permission(sd, static_cast<e_pc_permission>(script_getnum(st, 2))))
```
Comment (aleos89) [inline src/map/script.cpp:27354]:
```suggestion
	if (pc_has_permission(sd, perm))
```

File: src/map/script.cpp [hunk 11]
```diff
@@ -27340,6 +27340,53 @@ BUILDIN_FUNC(setdialogpospercent){
...
+ * removeperm(<permission>{,<char_id>})
+ */
+BUILDIN_FUNC(addperm)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(4, sd))
+		return SCRIPT_CMD_FAILURE;
+
+	e_pc_permission permission;
```
Comment (aleos89) [inline src/map/script.cpp:27374]: Another permission validation check, might be worth while to be a function!

File: src/map/script.cpp [hunk 12]
```diff
@@ -28105,6 +28152,10 @@ struct script_function buildin_func[] = {
 	BUILDIN_DEF(setdialogpos, "ii"),
 	BUILDIN_DEF(setdialogpospercent, "ii"),
 
+	BUILDIN_DEF(checkperm, "i?"),
```
Comment (aleos89) [inline src/map/script.cpp:28155]: Should we add aliases?
IE: "checkpermission", "addpermission", "removepermission"

**Review by aleos89 — APPROVED** [review]

File: src/map/script.cpp [hunk 13]
```diff
@@ -27340,6 +27340,56 @@ BUILDIN_FUNC(setdialogpospercent){
...
+ * permission_check(<permission>{,<char_id>}) -> <bool>
+ */
+BUILDIN_FUNC(permission_check)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(3, sd))
+		return SCRIPT_CMD_FAILURE;
+
+	e_pc_permission permission = static_cast<e_pc_permission>(script_getnum(st, 2));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27354]: You cast before verifying the range => undefined behavior for values that are not part of the enum.

File: src/map/script.cpp [hunk 14]
```diff
@@ -27340,6 +27340,56 @@ BUILDIN_FUNC(setdialogpospercent){
...
+ * permission_remove(<permission>{,<char_id>})
+ */
+BUILDIN_FUNC(permission_add)
+{
+	map_session_data* sd = nullptr;
+
+	if (!script_charid2sd(3, sd))
+		return SCRIPT_CMD_FAILURE;
+
+	e_pc_permission permission = static_cast<e_pc_permission>(script_getnum(st, 2));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27378]: You cast before verifying the range => undefined behavior for values that are not part of the enum.

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**